CMakeLists.txt

```diff
@@ -1,187 +1,311 @@
+# ------------------------------------------------------------------------------#
 # Options
-#------------------------------------------------------------------------------#
-option (SINGULARITY_USE_HDF5 "Pull in hdf5" ON)
-option (SINGULARITY_USE_FORTRAN "Enable fortran bindings" ON)
-option (SINGULARITY_USE_KOKKOS "Use Kokkos for portability" OFF)
-option (SINGULARITY_USE_EOSPAC "Pull in eospac" OFF)
-option (SINGULARITY_USE_CUDA "Enable cuda support" OFF)
-option (SINGULARITY_USE_KOKKOSKERNELS
-  "Use kokkos-kernels for linear algebra" OFF)
-option (SINGULARITY_BUILD_PYTHON "Compile Python bindings" OFF)
-option (SINGULARITY_BUILD_TESTS "Compile tests" OFF)
-option (SINGULARITY_BUILD_EXAMPLES "Compile examples" OFF)
-option (SINGULARITY_BUILD_SESAME2SPINER "Compile sesame2spiner" OFF)
-option (SINGULARITY_BUILD_STELLARCOLLAPSE2SPINER "Compile stellarcollapse2spiner" OFF)
-option (SINGULARITY_BETTER_DEBUG_FLAGS
-  "Better debug flags for singularity" ON)
-option (SINGULARITY_HIDE_MORE_WARNINGS "hide more warnings" OFF)
-option (SINGULARITY_SUBMODULE_MODE "Submodule mode" OFF)
-option (SINGULARITY_BUILD_CLOSURE "Mixed cell closure" ON)
-option (SINGULARITY_TEST_SESAME "Test the Sesame table readers" OFF)
-option (SINGULARITY_TEST_STELLAR_COLLAPSE "Test the stellar collapse table readers" OFF)
-option (SINGULARITY_TEST_PYTHON "Test the Python bindings" OFF)
-option (SINGULARITY_USE_SINGLE_LOGS "Use single precision logs. Can harm accuracy." OFF)
-option (SINGULARITY_USE_TRUE_LOG_GRIDDING "Use grids that conform to log spacing." OFF)
-option (SINGULARITY_PATCH_MPARK_VARIANT "Apply GPU patch to mpark-variant submodule" ON)
-
-# Patches variant to be compatible with cuda
-# Assumes "patch" is present on system
-if(SINGULARITY_PATCH_MPARK_VARIANT)
-  message(STATUS "Patching mpark::variant to support GPUs")
-  execute_process(COMMAND patch -N -s -V never
-                  ${CMAKE_CURRENT_SOURCE_DIR}/utils/variant/include/mpark/variant.hpp
-                  ${CMAKE_CURRENT_SOURCE_DIR}/utils/cuda_compatibility.patch
-                  )
-endif()
+# ------------------------------------------------------------------------------#
+include(CMakeDependentOption)
+
+# dependcy options
+option(SINGULARITY_USE_SPINER "Use spiner EOS" OFF)
+cmake_dependent_option(
+  SINGULARITY_USE_SPINER_WITH_HDF5 "Use HDF5 component of spiner" ON
+  "SINGULARITY_USE_SPINER" OFF)
+
+option(SINGULARITY_USE_FORTRAN "Enable fortran bindings" ON)
+option(SINGULARITY_USE_KOKKOS "Use Kokkos for portability" OFF)
+option(SINGULARITY_USE_EOSPAC "Pull in eospac" OFF)
+
+cmake_dependent_option(SINGULARITY_USE_CUDA "Use CUDA backend of Kokkos" ON
+                       "SINGULARITY_USE_KOKKOS" OFF)
+cmake_dependent_option(
+  SINGULARITY_USE_KOKKOSKERNELS "Use KokkosKernels for LA routines" ON
+  "SINGULARITY_USE_KOKKOS" OFF)
+cmake_dependent_option(
+  SINGULARITY_BUILD_CLOSURE "Mixed Cell Closure" ON
+  "SINGULARITY_USE_KOKKOSKERNELS;SINGULARITY_USE_KOKKOS" OFF)
```

Comment:
I think for future SAP work we'll want mixed cell closures (for hybrid EOS) but definitely _without_ the Kokkos/Kokkos Kernels dependency.

The main thing is that if you have closures on, you need some sort of linear solver compatible with the chosen portability strategy; you need Kokkos Kernels for Kokkos and you need Eigen for a `None` portability strategy

Reply:
Now that I think about it---where do the kokkos and cuda flags appear anymore? Since ports-of-call handles the portability strategy, is importing kokkos kernels the only place? There might be some opportunity for simplification there.

Regarding mixed cell closures without kokkos kernels, on GPU, that is currently hard to achieve because closures require an arbitrarily sized linear solver. I was imagining that on a first pass for the hybrid EOS, we would do a closure by hand which didn't require an arbitrarily sized matrix solve?

Reply:
The kokkos flags live here https://github.com/lanl/singularity-eos/blob/main/cmake/submodule_configs.cmake @Yurlungur. 

Other question, does this disable the ability to use Eigen for the solve @mauneyc-LANL ?

Reply:
To hopefully wrap up this tangent:

> I was imagining that on a first pass for the hybrid EOS, we would do a closure by hand which didn't require an arbitrarily sized matrix solve?

Yes I agree.

Reply:
Thanks @dholladay00 yeah that's right---the flags for kokkos kernels are actually pretty extensive. 

Regarding disabling Eigen: Are there currently any use-cases where Eigen would be used? SAP? It may be that we can safely remove Eigen but if so we should be deliberate about it.

Reply:
Eigen is used when `+closure~kokkos-kernels` in spack parlance (it may actually be `+closure~kokkos` as I'm not sure if we allow `+Eigen+kokkos`)

Reply:
Right---I know but do we ever want that?

Reply:
I don't think so...

Reply:
Probably we should keep it around until we have a little bit more clarity on what the FORTRAN codes will want. If they don't want us using Kokkos maybe we should hold on to Eigen for now (i.e., let's make sure it still works in this PR). But I think we can probably remove it in the near future as we figure out the plan for closures on device in those codes.

Reply:
I could foresee @Yurlungur lamenting this decision someday b/c kokkos is easy to add for quick development but kernels is not. When future @Yurlungur is angry with this decision and out of patience b/c spack and dependencies are hard and some new machine is also hard, remember this conversation and know that my ability to sympathize with the frustration will be 0. 

That being said, I'm ok either way. If we want to allow kokkos-serial+eigen or no mixing at all, either way I am ok.

Reply:
@dholladay00 what are you suggesting is the lower-pain path? Keeping Eigen? I am perfectly willing to be convinced to keep at around to avoid my own suffering

Reply:
Latest commit to address this

```cmake
option(SINGULARITY_BUILD_CLOSURE "Mixed Cell Closure" ON)
# ...
cmake_dependent_option(
  SINGULARITY_USE_KOKKOSKERNELS "Use KokkosKernels for LA routines" ON
  "SINGULARITY_USE_KOKKOS;SINGULARITY_BUILD_CLOSURE" OFF)
# ...
if(SINGULARITY_BUILD_CLOSURE AND NOT SINGULARITY_USE_KOKKOSKERNELS)
  set(SINGULARITY_USE_EIGEN
      ON
      CACHE BOOL "" FORCE)
else()
  set(SINGULARITY_USE_EIGEN
      OFF
      CACHE BOOL "" FORCE)
endif()
```

Reply:
> Latest commit to address this
> 
> ```cmake
> option(SINGULARITY_BUILD_CLOSURE "Mixed Cell Closure" ON)
> # ...
> cmake_dependent_option(
>   SINGULARITY_USE_KOKKOSKERNELS "Use KokkosKernels for LA routines" ON
>   "SINGULARITY_USE_KOKKOS;SINGULARITY_BUILD_CLOSURE" OFF)
> # ...
> if(SINGULARITY_BUILD_CLOSURE AND NOT SINGULARITY_USE_KOKKOSKERNELS)
>   set(SINGULARITY_USE_EIGEN
>       ON
>       CACHE BOOL "" FORCE)
> else()
>   set(SINGULARITY_USE_EIGEN
>       OFF
>       CACHE BOOL "" FORCE)
> endif()
> ```

That seems to make sense to me!

Reply:
Unfortunately we need to add in a guard against mixing this with cuda: if use cuda and build closure and not use kokkos kernels, then error.

Reply:
probably fine to just add a line in the first if clause to error out if cuda is enabled

Reply:
I've added a check for cuda/closure/kk

```cmake
# Use to determine if Eigen is used or not 
set(SINGULARITY_USE_EIGEN
      OFF
      CACHE BOOL "" FORCE)
if(SINGULARITY_BUILD_CLOSURE AND NOT SINGULARITY_USE_KOKKOSKERNELS)
  if(SINGULARITY_USE_CUDA)
    message(FATAL_ERROR "\"SINGULARITY_BUILD_CLOSURE=ON\" and \"SINGULARITY_USE_CUDA=ON\" requires \"SINGULARITY_USE_KOKKOSKERNELS=ON\"")
  set(SINGULARITY_USE_EIGEN
      ON
      CACHE BOOL "" FORCE)
endif()
```